services/sso/src/main/java/software.amazon.awssdk.services.sso/auth/SsoCredentialsProvider.java

```diff
@@ -0,0 +1,252 @@
+    private final Duration prefetchTime;
+
+    private final CachedSupplier<SessionCredentialsHolder> credentialCache;
+
+    /**
+     * @see #builder()
+     */
+    private SsoCredentialsProvider(BuilderImpl builder) {
+        this.ssoClient = notNull(builder.ssoClient, "SSO client must not be null.");
+        this.getRoleCredentialsRequestSupplier = builder.getRoleCredentialsRequestSupplier;
+
+        this.staleTime = Optional.ofNullable(builder.staleTime).orElse(DEFAULT_STALE_TIME);
+        this.prefetchTime = Optional.ofNullable(builder.prefetchTime).orElse(DEFAULT_PREFETCH_TIME);
+
+        CachedSupplier.Builder<SessionCredentialsHolder> cacheBuilder = CachedSupplier.builder(this::updateSsoCredentials);
+        if (builder.asyncCredentialUpdateEnabled) {
+            cacheBuilder.prefetchStrategy(new NonBlocking(ASYNC_THREAD_NAME));
+        }
+
+        this.credentialCache = cacheBuilder.build();
+    }
+
+    /**
+     * Update the expiring session SSO credentials by calling SSO. Invoked by {@link CachedSupplier} when the credentials
+     * are close to expiring.
+     */
+    private RefreshResult<SessionCredentialsHolder> updateSsoCredentials() {
+        SessionCredentialsHolder credentials = getUpdatedCredentials(ssoClient);
+        Instant acutalTokenExpiration = credentials.getSessionCredentialsExpiration();
+
+        return RefreshResult.builder(credentials)
+                            .staleTime(acutalTokenExpiration.minus(staleTime))
+                            .prefetchTime(acutalTokenExpiration.minus(prefetchTime))
+                            .build();
+    }
+
+    private SessionCredentialsHolder getUpdatedCredentials(SsoClient ssoClient) {
+        GetRoleCredentialsRequest request = getRoleCredentialsRequestSupplier.get();
+        notNull(request, "GetRoleCredentialsRequest can't be null.");
+        RoleCredentials roleCredentials = ssoClient.getRoleCredentials(request).roleCredentials();
+        AwsSessionCredentials sessionCredentials = AwsSessionCredentials.create(roleCredentials.accessKeyId(),
+                                                                                roleCredentials.secretAccessKey(),
+                                                                                roleCredentials.sessionToken());
+        return new SessionCredentialsHolder(sessionCredentials, Instant.ofEpochMilli(roleCredentials.expiration()));
+    }
+
+    /**
+     * The amount of time, relative to session token expiration, that the cached credentials are considered stale and
+     * should no longer be used. All threads will block until the value is updated.
+     */
+    public Duration staleTime() {
+        return staleTime;
+    }
+
+    /**
+     * The amount of time, relative to session token expiration, that the cached credentials are considered close to stale
+     * and should be updated.
+     */
+    public Duration prefetchTime() {
+        return prefetchTime;
+    }
```

Comment:
Do these need to be public?

Reply:
Not really for now, I made them public for consistency with the existed STSCredentialsProvider, I can change it.

Reply:
Ah, if StsCredentialsProvider have them public I guess that's fine.